packages/core/src/core/plan/plan.ts

```diff
@@ -39,33 +38,93 @@ const { resolve, resolveReferenceExpression } = expressions
 
 const log = logger(module)
 
+const COMPARE_RETURN = 'return'
+const COMPARE_RECURSE = 'recurse'
+
+type CompareReturnValue = {
+  returnCode: typeof COMPARE_RETURN
+  returnValue: boolean
+} | {
+  returnCode: typeof COMPARE_RECURSE
+  returnValue: {
+    firstValue: Value
+    secondValue: Value
+  }
+}
```

Comment:
trust the typescript compiler, there is no need to define constants when all you need to check is in the code and not at runtime
Also, there is a lot of "compare" code, the type name should probably be more specific
```suggestion
type ReferenceCompareReturnValue = {
  returnCode: 'return'
  returnValue: boolean
} | {
  returnCode: 'recurse'
  returnValue: {
    firstValue: Value
    secondValue: Value
  }
}
```